crates/subspace-networking/src/utils/rate_limiter.rs

```diff
@@ -30,45 +31,104 @@ const REGULAR_BASE_CONCURRENT_TASKS: NonZeroUsize =
 /// to be tweaked in the future.
 pub(crate) const REGULAR_CONCURRENT_TASKS_BOOST_PER_PEER: usize = 25;
 
+/// Defines the minimum size of the "connection limit semaphore".
+const MINIMUM_CONNECTIONS_SEMAPHORE_SIZE: usize = 3;
+
+#[derive(Debug)]
+pub(crate) struct RateLimiterPermit {
+    /// Limits Kademlia substreams.
+    _substream_limit_permit: ResizableSemaphorePermit,
+
+    /// Limits outgoing connections.
+    _connection_limit_permit: ResizableSemaphorePermit,
+}
+
 #[derive(Debug)]
 pub(crate) struct RateLimiter {
     kademlia_tasks_semaphore: ResizableSemaphore,
     regular_tasks_semaphore: ResizableSemaphore,
+    connections_semaphore: ResizableSemaphore,
 }
 
 impl RateLimiter {
-    pub(crate) fn new() -> Self {
+    pub(crate) fn new(out_connections: u32, pending_out_connections: u32) -> Self {
+        let permits = Self::calculate_connection_semaphore_size(
+            out_connections as usize,
+            pending_out_connections as usize,
+        );
+
+        debug!(%out_connections, %pending_out_connections, %permits, "Rate limiter was instantiated.");
+
         Self {
             kademlia_tasks_semaphore: ResizableSemaphore::new(KADEMLIA_BASE_CONCURRENT_TASKS),
             regular_tasks_semaphore: ResizableSemaphore::new(REGULAR_BASE_CONCURRENT_TASKS),
+            connections_semaphore: ResizableSemaphore::new(
+                NonZeroUsize::new(permits).expect("Manual setting."),
```

Comment:
This is a confusing proof to be completely honest. And considering that `Self::calculate_connection_semaphore_size` literally returns `0` in the very first branch it is also incorrect. If you can guarantee `Self::calculate_connection_semaphore_size` to never return zero, then return `NonZeroUsize` from it instead of constructing it here.